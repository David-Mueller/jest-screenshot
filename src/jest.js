"use strict";
exports.__esModule = true;
function isSnapshotState(obj) {
    if (typeof obj !== "object") {
        return false;
    }
    var _counters = obj._counters;
    if (typeof _counters !== "object") {
        return false;
    }
    if (typeof _counters.get !== "function") {
        return false;
    }
    return true;
}
exports.isSnapshotState = isSnapshotState;
function isJestTestConfiguration(obj) {
    if (typeof obj !== "object") {
        return false;
    }
    var snapshotState = obj.snapshotState, testPath = obj.testPath, currentTestName = obj.currentTestName, isNot = obj.isNot;
    if (typeof testPath !== "string") {
        return false;
    }
    if (typeof currentTestName !== "string") {
        return false;
    }
    if (typeof isNot !== "boolean") {
        return false;
    }
    if (!isSnapshotState(snapshotState)) {
        return false;
    }
    return true;
}
exports.isJestTestConfiguration = isJestTestConfiguration;
