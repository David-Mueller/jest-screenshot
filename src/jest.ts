export interface SnapshotState {
    _counters: Map<string, number>;
    _updateSnapshot: "new" | "all" | "none";
}

export interface JestTestConfiguration {
    snapshotState: SnapshotState;
    testPath: string;
    currentTestName: string;
    isNot: boolean;
}

export interface MatcherResult {
    message?(): string;
    pass: boolean;
}

export function isSnapshotState(obj: any): obj is SnapshotState {
    if (typeof obj !== "object") { return false; }
    const { _counters } = obj;
    if (typeof _counters !== "object") { return false; }
    if (typeof _counters.get !== "function") { return false; }
    return true;
}

export function isJestTestConfiguration(obj: any): obj is JestTestConfiguration {
    if (typeof obj !== "object") { return false; }
    const { snapshotState, testPath, currentTestName, isNot } = obj;
    if (typeof testPath !== "string") { return false; }
    if (typeof currentTestName !== "string") { return false; }
    if (typeof isNot !== "boolean") { return false; }
    if (!isSnapshotState(snapshotState)) { return false; }
    return true;
}
