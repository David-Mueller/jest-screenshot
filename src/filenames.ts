import { SnapshotState } from "./jest";
import { ToMatchImageSnapshotConfiguration } from "./to-match-image-snapshot";
import * as path from "path";
import kebabCase from "lodash.kebabcase";

export function getSnapshotFileName(
    testPath: string,
    currentTestName: string,
    snapshotState: SnapshotState,
    { identifier }: ToMatchImageSnapshotConfiguration,
) {
    const counter = (snapshotState._counters.get(currentTestName) || 0) + 1;
    if (typeof identifier === "function") {
        const fileName = identifier(testPath, currentTestName, counter);
        if (fileName.toLowerCase().endsWith(".snap.png")) {
            return fileName;
        }
        return `${fileName}.snap.png`;
    }
    if (typeof identifier !== "undefined") {
        throw new Error("Jest: Invalid configuration for `.toMatchImageSnapshot`: `identifier` must be a function.");
    }
    return `${kebabCase(path.basename(testPath))}-${kebabCase(currentTestName)}-${counter}.snap.png`;
}

export function getSnapshotPath(
    testPath: string,
    currentTestName: string,
    snapshotState: SnapshotState,
    configuration: ToMatchImageSnapshotConfiguration,
) {
    const { snapshotsDir } = configuration;
    const fileName = getSnapshotFileName(testPath, currentTestName, snapshotState, configuration);
    return path.join(path.dirname(testPath), snapshotsDir || "__snapshots__", fileName);
}
