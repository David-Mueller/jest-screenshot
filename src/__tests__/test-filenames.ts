import { getSnapshotFileName, getSnapshotPath } from "../filenames";
import { SnapshotState } from "../jest";

const somePath = "/tmp/some-test.ts";
const someTestName = "something with some parameters does expected things";
const someSnapshotState: SnapshotState = {
    _counters: new Map<string, number>(),
    _updateSnapshot: "new",
    updated: 0,
    added: 0,
};
const expectedGeneratedName = "some-test-ts-something-with-some-parameters-does-expected-things-8.snap.png";
someSnapshotState._counters.set(someTestName, 7);

describe("getSnapshotFileName", () => {
    describe("with a custom identifier generator", () => {
        it("uses the generator", () => {
            const expected = "some-generated-filename-identifier.snap.png";
            const identifier = (testPath: string, currentTestName: string, counter: number) => {
                expect(testPath).toBe(somePath);
                expect(currentTestName).toBe(someTestName);
                expect(counter).toBe(8);
                return expected;
            };
            expect(getSnapshotFileName(somePath, someTestName, someSnapshotState, { identifier })).toBe(expected);
        });

        it("appends `.snap.png` if the generator didn't append it", () => {
            const expected = "some-generated-filename-identifier.snap.png";
            const returned = "some-generated-filename-identifier";
            const identifier = () => {
                return returned;
            };
            expect(getSnapshotFileName(somePath, someTestName, someSnapshotState, { identifier })).toBe(expected);
        });

        it("throws an error if the generator wasn't a function but also not `undefined`", () => {
            const identifier = 89 as any;
            expect(() => {
                getSnapshotFileName(somePath, someTestName, someSnapshotState, { identifier });
            }).toThrowErrorMatchingSnapshot();
        });
    });

    it("generates the expected default file name with no generator specified", () => {
        expect(getSnapshotFileName(somePath, someTestName, someSnapshotState, {})).toBe(expectedGeneratedName);
    });
});

describe("getSnapshotPath", () => {
    it("returns the expected path", () => {
        const expectedGeneratedPath = `/tmp/__snapshots__/${expectedGeneratedName}`;
        expect(getSnapshotPath(somePath, someTestName, someSnapshotState, {})).toBe(expectedGeneratedPath);
    });

    it("returns the expected path with a custom directory specified", () => {
        const snapshotsDir = "__some_directory__";
        const expectedGeneratedPath = `/tmp/__some_directory__/${expectedGeneratedName}`;
        const path = getSnapshotPath(somePath, someTestName, someSnapshotState, { snapshotsDir });
        expect(path).toBe(expectedGeneratedPath);
    });
});
