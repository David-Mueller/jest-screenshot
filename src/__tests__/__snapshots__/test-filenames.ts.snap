// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`getSnapshotFileName with a custom identifier generator throws an error if the generator wasn't a function but also not \`undefined\` 1`] = `"Jest: Invalid configuration for \`.toMatchImageSnapshot\`: \`identifier\` must be a function."`;
